### PR TITLE
Feature/celebrity-birthdays

### PR DESCRIPTION
**Pull request by advaa123**
Optional feature

celeb

### CODE REVIEW

File: app/internal/celebrity.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    except requests.ConnectionError:
+        # TODO: write error to log file
+        return None
+    else:
+        if page.status_code == 200:
+            return page.content
+        return None
+
+
+def get_celebs(today: str) -> Optional[Dict]:
```
Comment (yammesicka) [inline app/internal/celebrity.py:32]: Expand the type annotation: Dict[key_type, value_type]

File: app/internal/celebrity.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    return datetime.date.today().strftime("%m-%d")
+
+
+def get_content(today: str) -> Optional[bytes]:
+    """Get IMDB HTML page content by today's birthdays"""
+
+    url = BASE_URL + today
+
+    try:
+        page = requests.get(url)
```
Comment (yammesicka) [inline app/internal/celebrity.py:22]: Prefer sending the request using JS

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: app/static/celebrity.css [hunk 3]
```diff
@@ -0,0 +1,15 @@
+.top-div {
+    display: block;
+    width: 180px;
```
Comment (yammesicka) [inline app/static/celebrity.css:3]: Prefer `em`

File: app/static/celebrity.css [hunk 4]
```diff
@@ -0,0 +1,15 @@
+.top-div {
+    display: block;
```
Comment (yammesicka) [inline app/static/celebrity.css:2]: Maybe `flex`?

File: app/static/celebrity.js [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+                for (let i = 0; i < 12; i++) {
+                    namesArray.push(names[i].innerText.trim().substr(5).trim());
+                }
+                return namesArray;
+            }
+
+            function getImages() {
+                let imagesArray = new Array();
+                const images = doc.querySelectorAll('div.lister-item-image a img');
+                for (let i = 0; i < 12; i++) {
```
Comment (yammesicka) [inline app/static/celebrity.js:24]: What is 12? Move to `const` with a proper name

File: app/static/celebrity.js [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            function getNames() {
+                let namesArray = new Array();
+                const names = doc.querySelectorAll('h3.lister-item-header');
+                for (let i = 0; i < 12; i++) {
+                    namesArray.push(names[i].innerText.trim().substr(5).trim());
+                }
+                return namesArray;
+            }
+
+            function getImages() {
```
Comment (yammesicka) [inline app/static/celebrity.js:21]: Looks like you can create a more abstract version of `getNames`, `getImages` and `getJobs`

File: app/static/celebrity.js [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+    fetch('https://www.imdb.com/search/name/?birth_monthday=' + dateToday)
+        .then(function (response) {
+            return response.text();
+        }).then(function (html) {
+            celebCards.innerHTML = "";
+            const parser = new DOMParser();
+            const doc = parser.parseFromString(html, 'text/html');
+
+            function getNames() {
```
Comment (yammesicka) [inline app/static/celebrity.js:12]: Move out to an unindented position.

File: tests/test_celebrity.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import datetime
+
+import pytest
+from app.internal.celebrity import get_today_month_and_day
+
+
+CELEBRITY_PAGE = "/celebrity"
+FAKE_TIME = datetime.date(2018, 9, 18)
```
Comment (yammesicka) [inline tests/test_celebrity.py:8]: Use pytest-hypothesis

**Review by yammesicka — CHANGES_REQUESTED** [review]
Great job! Added some insights

File: app/static/celebrity.js [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    fetch('https://www.imdb.com/search/name/?birth_monthday=' + dateToday)
+        .then(function (response) {
+            return response.text();
+        }).then(function (html) {
+            celebCards.innerHTML = "";
+            const parser = new DOMParser();
+            const doc = parser.parseFromString(html, 'text/html');
+            getCelebsData(doc);
+        }).catch(() => {
+            celebCards.innerHTML = "Failed to load Celebrity birthdays."
```
Comment (yammesicka) [inline app/static/celebrity.js:14]: `;`?

File: app/static/celebrity.js [hunk 10]
```diff
@@ -0,0 +1,101 @@
+function getCelebs(dateToday) {
+    const celebCards = document.getElementById("celebCards");
+    const numberOfCelebs = 12;
+
+    fetch('https://www.imdb.com/search/name/?birth_monthday=' + dateToday)
+        .then(function (response) {
+            return response.text();
+        }).then(function (html) {
+            celebCards.innerHTML = "";
```
Comment (yammesicka) [inline app/static/celebrity.js:9]: Prefer innerText when possible

File: app/static/celebrity.js [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        }).catch(() => {
+            celebCards.innerHTML = "Failed to load Celebrity birthdays."
+        });
+
+    function getCelebsData(doc) {
+        function getArray(selector, dataType) {
+            let myArray = new Array();
+            const selectorData = doc.querySelectorAll(selector);
+            for (let i = 0; i < numberOfCelebs; i++) {
+                if (dataType === "names") {
```
Comment (yammesicka) [inline app/static/celebrity.js:22]: Split to a different function

File: app/static/celebrity.js [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+                card.appendChild(divFooter);
+                celebCards.appendChild(card);
+            }
+        }
+
+        const namesArray = getArray('h3.lister-item-header', "names");
+        const imagesArray = getArray('div.lister-item-image a img', "images");
+        const linksArray = getArray('div.lister-item-image a', "links");
+        const jobsArray = getArray("p.text-muted", "jobs");
```
Comment (yammesicka) [inline app/static/celebrity.js:98]: Choose one: ' or "

File: app/templates/celebrity.html [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+        <div id="celebCards" class="card-deck" style="flex-flow: row wrap; display: flex;">
+            <div class="spinner-border" role="status">
+                <span class="sr-only"></span>
+              </div>
+        </div>
+    </div>
+</div>
+<link rel="stylesheet" href="{{ url_for('static', path='/celebrity.css') }}">
+<script type="text/javascript" src="{{ url_for('static', path='/celebrity.js')}}"></script>
+<script>
```
Comment (yammesicka) [inline app/templates/celebrity.html:19]: Move to the js file :)

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: app/templates/celebrity.html (hunk 13 above)
Comment (advaa123) [reply app/templates/celebrity.html:19]: I'm passing a Python customized date variable (**{{date | safe}}**) from the Jinja2 template to the JS file, and I can't do that directly from the JS file (also, I find this method very friendly). How do you suggest doing so without adding more code? Thank you in advance!

File: tests/test_celebrity.py [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+
+
+CELEBRITY_PAGE = "/celebrity"
+DATE_START = datetime.date(2019, 1, 1)
+DATE_END = datetime.date(2022, 1, 1)
+
+
+@given(fake_date=st.dates(DATE_START, DATE_END))
+def test_get_today_month_and_day(fake_date):
+    if datetime.date.today().strftime("%m-%d") == fake_date.strftime("%m-%d"):
```
Comment (yammesicka) [inline tests/test_celebrity.py:15]: `datetime.date.today().strftime("%m-%d") == get_today_month_and_day()`?

Isn't that what the function actually returns? ^^"

File: app/static/celebrity.js [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+        const namesArray = getArray("h3.lister-item-header", "names");
+        const imagesArray = getArray("div.lister-item-image a img", "images");
+        const linksArray = getArray("div.lister-item-image a", "links");
+        const jobsArray = getArray("p.text-muted", "jobs");
+        buildCards(celebCards);
+    }
+}
+
+getCelebs(celebCards.dataset.today);
```
Comment (yammesicka) [inline app/static/celebrity.js:107]: What for the entire page to load, first :)

**Review by yammesicka — CHANGES_REQUESTED** [review]
Great work so far :)

File: tests/test_celebrity.py (hunk 14 above)
Comment (advaa123) [reply tests/test_celebrity.py:15]: Haha help please, I don’t understand how to test it then

File: app/static/celebrity.js (hunk 15 above)
Comment (advaa123) [reply app/static/celebrity.js:107]: Like this kind of thing?
window.addEventListener('load', function () {
  getCelebs(....);
});

Comment (yammesicka) [reply app/static/celebrity.js:107]: Yeah :)

**Review by yammesicka — APPROVED** [review]

**Pull request merged**